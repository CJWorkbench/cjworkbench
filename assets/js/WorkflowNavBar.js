import React from 'react'
import PropTypes from 'prop-types'
import WfHamburgerMenu from './WfHamburgerMenu'
import UndoRedoButtons from './UndoRedoButtons'
import ConnectedEditableWorkflowName, { EditableWorkflowName } from './EditableWorkflowName'
import { goToUrl, timeDifference } from './utils'
import ShareButton from './ShareModal/ShareButton'
import { Trans, t } from '@lingui/macro'
import { withI18n } from '@lingui/react'

function NoOp () {}

function LessonCourse ({ localeId, course }) {
  let path
  let title

  if (course) {
    path = `/courses/${course.localeId}/${course.slug}`
    title = course.title
  } else {
    path = `/lessons/${localeId}`
    title = <Trans id='js.WorkflowNavBar.LessonCourse.Lesson.title'>Training</Trans>
  }

  return (
    <div className='course'>
      <a href={path}>{title}</a>
    </div>
  )
}

function LessonWorkflowTitle ({ lesson }) {
  return (
    <div className='title-metadata-stack'>
      <LessonCourse localeId={lesson.localeId} course={lesson.course} />
      <EditableWorkflowName
        value={lesson.header.title}
        setWorkflowName={NoOp}
        isReadOnly
      />
    </div>
  )
}

const OwnedWorkflowTitleAndMetadata = withI18n()(function ({ i18n, isReadOnly, workflow }) {
  const owner = workflow.owner_name.trim()
  const timeAgo = timeDifference(workflow.last_update, new Date(), i18n)
  return (
    <div className='title-metadata-stack'>
      <ConnectedEditableWorkflowName isReadOnly={isReadOnly} />
      <ul className='metadata-container'>
        {!workflow.is_anonymous ? (
          <li className='attribution'>
            <span className='metadata'><Trans id='js.WorkflowNavBar.OwnedWorkflowTitleAndMetadata.owner'>by {owner}</Trans> </span>
            <span className='separator'>-</span>
          </li>
        ) : null}
        <li>
          <Trans id='js.WorkflowNavBar.OwnedWorkflowTitleAndMetadata.lastUpdated' description="{timeAgo} will contain something like '4h ago'">
            Updated {timeAgo}
          </Trans>
        </li>
        {(!isReadOnly && !workflow.is_anonymous) ? (
          <li>
            <span className='separator'>-</span>
            <ShareButton>
              {workflow.public ? <Trans id='js.WorkflowNavBar.OwnedWorkflowTitleAndMetadata.visibility.public'>Public</Trans> : <Trans id='js.WorkflowNavBar.OwnedWorkflowTitleAndMetadata.visibility.private'>Private</Trans>}
            </ShareButton>
          </li>
        ) : null}
      </ul>
    </div>
  )
})

function WorkflowTitleAndMetadata ({ lesson, isReadOnly, workflow }) {
  if (lesson) {
    return (
      <LessonWorkflowTitle
        lesson={lesson}
      />
    )
  } else {
    return (
      <OwnedWorkflowTitleAndMetadata
        isReadOnly={isReadOnly}
        workflow={workflow}
      />
    )
  }
}

export class WorkflowNavBar extends React.Component {
  static propTypes = {
    i18n: PropTypes.shape({
      // i18n object injected by LinguiJS withI18n()
      _: PropTypes.func.isRequired
    }),
    api: PropTypes.object.isRequired,
    workflow: PropTypes.object.isRequired,
    lesson: PropTypes.shape({
      course: PropTypes.shape({
        slug: PropTypes.string.isRequired,
        title: PropTypes.string.isRequired
      }), // optional -- no course means plain lesson
      header: PropTypes.shape({
        title: PropTypes.string.isRequired
      }).isRequired
    }), // optional -- no lesson means we're not in the "lessons" interface
    isReadOnly: PropTypes.bool.isRequired,
    loggedInUser: PropTypes.object // undefined if no user logged in
  }

  state = {
    spinnerVisible: false,
    isShareModalOpen: false
  }

  componentWillUnmount = () => {
    this.unmounted = true
  }

  undoRedo (verb) {
    // TODO use reducer for this, with a global "can't tell what's going to
    // change" flag instead of this.state.spinnerVisible.

    // Prevent keyboard shortcuts or mouse double-undoing.
    if (this.state.spinnerVisible) return

    this.setState({ spinnerVisible: true })
    this.props.api[verb](this.props.workflow.id)
      .then(() => {
        if (this.unmounted) return
        this.setState({ spinnerVisible: false })
      })
  }

  undo = () => {
    this.undoRedo('undo')
  }

  redo = () => {
    this.undoRedo('redo')
  }

  handleDuplicate = () => {
    if (!this.props.loggedInUser) {
      // user is NOT logged in, so navigate to sign in
      goToUrl('/account/login')
    } else {
      // user IS logged in: start spinner, make duplicate & navigate there
      this.setState({ spinnerVisible: true })

      this.props.api.duplicateWorkflow(this.props.workflow.id)
        .then(json => {
          goToUrl('/workflows/' + json.id)
        })
    }
  }

  render () {
    const { api, isReadOnly, loggedInUser, lesson, workflow, i18n } = this.props

    // menu only if there is a logged-in user
    let contextMenu
    if (loggedInUser) {
      contextMenu = (
        <WfHamburgerMenu
          workflowId={workflow.id}
          api={api}
          isReadOnly={isReadOnly}
          user={loggedInUser}
        />
      )
    } else {
      contextMenu = (
        <a href='/account/login' className='nav--link'>{i18n._(t('js.WorkflowNavBar.signIn.accountLink')`Sign in`)}</a>
      )
    }

    const spinner = this.state.spinnerVisible ? (
      <div className='spinner-container'>
        <div className='spinner-l1'>
          <div className='spinner-l2'>
            <div className='spinner-l3' />
          </div>
        </div>
      </div>
    ) : null

    return (
      <>
        {spinner}
        <nav className='navbar'>
          <div className='navbar-elements'>
            <a href='/workflows/' className='logo-navbar'>
              <img className='image' src={`${window.STATIC_URL}images/logo.svg`} />
            </a>
            <WorkflowTitleAndMetadata
              lesson={lesson}
              isReadOnly={isReadOnly}
              workflow={workflow}
            />
            <div className='nav-buttons'>
              {isReadOnly ? null : (
                <UndoRedoButtons undo={this.undo} redo={this.redo} />
              )}
              <button name='duplicate' onClick={this.handleDuplicate}>{i18n._(t('js.WorkflowNavBar.duplicate.button')`Duplicate`)}</button>
              {lesson ? null : (/* We haven't yet designed what it means to share a lesson workflow. */
                <ShareButton>{i18n._(t('js.WorkflowNavBar.share.shareButton')`Share`)}</ShareButton>
              )}
              {contextMenu}
            </div>
          </div>
        </nav>
      </>
    )
  }
}
export default withI18n()(WorkflowNavBar)
