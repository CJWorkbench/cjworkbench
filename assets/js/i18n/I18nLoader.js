import React from 'react'
import { I18nProvider } from '@lingui/react'
import fetchCatalog from './catalogs'

export class I18nLoader extends React.Component {
  state = {
    catalogs: {},
  }
  
  loadCatalog = (language) => {
      this.setState(state => ({
          catalogs: {
            ...state.catalogs,
            [language]: fetchCatalog(language)
          }
      }))
  }
  
  componentDidMount() {
    this.loadCatalog(this.props.language)
  }

  shouldComponentUpdate(nextProps, nextState) {
    const { language } = nextProps
    const { catalogs } = nextState

    if (language !== this.props.language && !catalogs[language]) {
      this.loadCatalog(language)
      return false
    }

    return true
  }

  render () {
    const { children, language } = this.props
    const { catalogs } = this.state

    // Skip rendering when catalog isn't loaded.
    if (!catalogs[language]) {
        return null
    }

    return (
      <I18nProvider language={language} catalogs={catalogs}>
        {children}
      </I18nProvider>
    )
  }
}
