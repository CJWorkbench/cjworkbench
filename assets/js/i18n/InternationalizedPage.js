import React from 'react'
import PropTypes from 'prop-types'
import { en, el } from 'make-plural/plurals'
import { i18n } from '@lingui/core'
import { I18nProvider } from '@lingui/react'
import { currentLocaleId } from './locales'
import Spinner from '../Spinner'

__webpack_public_path__ = window.STATIC_URL + 'bundles/' // eslint-disable-line

async function loadLocale (localeId) {
  const { messages } = await import(
    /* webpackChunkName: "[request]" */
    `@lingui/loader!../../locale/${localeId}/messages.po`
  )
  i18n.loadLocaleData({
    en: { plurals: en },
    el: { plurals: el }
  })
  i18n.load(localeId, messages)
  i18n.activate(localeId)
  return i18n
}

function useActivatedI18n (localeId) {
  const [i18n, setI18n] = React.useState(null)

  React.useEffect(() => {
    loadLocale(currentLocaleId).then(setI18n)
  }, [])

  return i18n
}

export default function InternationalizedPage (props) {
  const { children } = props
  const i18n = useActivatedI18n(currentLocaleId)

  if (i18n === null) {
    return <Spinner />
  }

  return (
    <I18nProvider i18n={i18n} forceRenderOnLocaleChange={false}>
      {children}
    </I18nProvider>
  )
}
InternationalizedPage.propTypes = {
  children: PropTypes.node.isRequired
}
