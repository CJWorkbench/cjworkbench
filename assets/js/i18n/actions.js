const SET_LOCALE = 'SET_LOCALE';

function reduceSetLocale(state, action){
  return {
    ...state,
    locale: action.locale
  };
}

export function setLocaleAction(locale){
  return {
    type: SET_LOCALE,
    locale: locale
  }
}

export const reducerFunctions = {
  [SET_LOCALE]: reduceSetLocale
}

export function localeReducer(state, action){
  if (action.type in reducerFunctions) {
    // Run a registered reducer
    return reducerFunctions[action.type](state, action)
  }
  
  return state;
}
