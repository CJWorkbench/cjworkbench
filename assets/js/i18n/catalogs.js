import enCatalog from '../../locale/en/messages.po'
import elCatalog from '../../locale/el/messages.po'

const catalogs = {
  en: enCatalog,
  el: elCatalog
}

export default function fetchCatalog (language) {
  return catalogs[language]
}
