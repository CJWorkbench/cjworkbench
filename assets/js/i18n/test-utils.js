import { I18nProvider } from '@lingui/react'
import React from 'react'
import { shape, object } from 'prop-types'
import { mount, shallow } from 'enzyme'

/*
 * Below lie utils for testing with i18n.
 * See https://lingui.js.org/guides/testing.html
 */
// Create the I18nProvider to retrieve context for wrapping around.
const intlProvider = new I18nProvider({
  language: 'en',
  catalogs: {
    en: {}
  }
}, {})

const {
  linguiPublisher: {
    i18n: originalI18n
  }
} = intlProvider.getChildContext()

// You customize the i18n object here:
export const i18n = {
  ...originalI18n,
  // provide _ macro, for trying to unwrap default/id or just passing down the key
  _: key => {
    if (typeof key.defaults === 'string') return key.defaults
    if (typeof key.id === 'string') return key.id
    return key
  }
}

/**
 * When using Lingui `withI18n` on components, props.i18n is required.
 */
function nodeWithI18nProp (node) {
  return React.cloneElement(node, { i18n })
}

/*
 * Methods to use
 */
export function shallowWithI18n (node, { context } = {}) {
  return shallow(
    nodeWithI18nProp(node),
    {
      context: Object.assign({}, context, { i18n })
    }
  )
}

export function mountWithI18n (node, { context, childContextTypes } = {}) {
  const newContext = Object.assign({}, context, { linguiPublisher: { i18n } })
  /*
   * I18nProvider sets the linguiPublisher in the context for withI18n to get
   * the i18n object from.
   */
  const newChildContextTypes = Object.assign({},
    {
      linguiPublisher: shape({
        i18n: object.isRequired
      }).isRequired
    },
    childContextTypes
  )
  return mount(
    nodeWithI18nProp(node),
    {
      context: newContext,
      childContextTypes: newChildContextTypes
    }
  )
}
