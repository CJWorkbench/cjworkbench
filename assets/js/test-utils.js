/* globals jest */
import { createStore, applyMiddleware } from 'redux'
import thunk from 'redux-thunk'
import promiseMiddleware from 'redux-promise-middleware'
import errorMiddleware from './error-middleware'
import { workflowReducer } from './workflow-reducer'

// Returns new mock function that returns given json. Used for mocking "get" API calls
export function jsonResponseMock (json) {
  return jest.fn().mockImplementation(() =>
    Promise.resolve(json)
  )
}

// Returns new mock function that gives an OK HTTP response. Use for mocking "set" API calls
export function okResponseMock () {
  return jsonResponseMock(null)
}

// Helper function to return a promise that resolves after all other promise mocks,
// even if they are chained like Promise.resolve().then(...).then(...)
// Technically: this is designed to resolve on the next macrotask
// https://stackoverflow.com/questions/25915634/difference-between-microtask-and-macrotask-within-an-event-loop-context
export function tick () {
  return new Promise(resolve => {
    setTimeout(resolve, 0)
  })
}

export function sleep (ms) {
  return new Promise(resolve => {
    setTimeout(resolve, ms)
  })
}

/**
 * Return `[ notify, promise ]`; `promise` completes when `notify` is called.
 *
 * Usage:
 *
 *     const [ notify, condition ] = createConditionVariable
 *     const done = condition.then(do_something)
 *     await Promise.all([sleep(0.1).then(notify), done])
 */
export function createConditionVariable () {
  let notify
  const promise = new Promise(resolve => { notify = resolve })
  return [notify, promise]
}

/**
 * Mock Redux store with in-production reducers.
 *
 * This is for integration-style tests: they test the reducer _and_ the caller
 * at the same time.
 *
 * Usage:
 *
 *     const api = { undo: jest.fn() }
 *     const store = mockStoreWithReducer({'workflow': ..., ...}, api)
 *     store.dispatch(Actions.undo())
 *     expect(api.undo).toHaveBeenCalled()
 *     expect(store.getState().workflow).toEqual(...)
 */
export function mockStore (initialState, api = {}) {
  const middlewares = [errorMiddleware(), promiseMiddleware, thunk.withExtraArgument(api)]
  const store = createStore(workflowReducer, initialState, applyMiddleware(...middlewares))
  return store
}
