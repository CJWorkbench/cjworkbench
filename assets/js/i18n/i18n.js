import { setupI18n } from "@lingui/core"
import { currentLocale } from './locales'
import { fetchCatalog } from './catalogs'

export default const i18n = setupI18n({
   language: currentLocale,
   catalogs: {
       currentLocale: fetchCatalog(currentLocale)
   }
})
