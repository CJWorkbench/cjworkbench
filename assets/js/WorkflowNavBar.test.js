/* globals beforeEach, describe, expect, it, jest */
import React from 'react'
import { WorkflowNavBar } from './WorkflowNavBar'
import { tick } from './test-utils'
import { shallowWithI18n, mountWithI18n } from './i18n/test-utils'

import Utils from './utils'

jest.mock('./utils', () => ({
  goToUrl: jest.fn()
}))

describe('WorkflowNavBar', () => {
  beforeEach(() => {
    Utils.goToUrl.mockReset()
  })

  it('should link back to /lessons when viewing a lesson', () => {
    const workflow = {
      id: 12,
      name: 'Original Version',
      owner_name: 'John Johnson',
      public: true
    }

    const lesson = {
      course: null,
      localeId: 'en',
      header: { title: 'A Lesson' }
    }

    const api = { duplicateWorkflow: jest.fn() }

    const wrapper = mountWithI18n(
      <WorkflowNavBar
        workflow={workflow}
        api={api}
        lesson={lesson}
        isReadOnly={false}
      /> // no loggedInUser prop
    )

    const a = wrapper.find('.course a')
    expect(a.prop('href')).toEqual('/lessons/en')
    expect(a.find('Trans').prop('defaults')).toEqual('Training') // hard-coded
  })

  it('should link back to /courses/slug when viewing a lesson in a course', () => {
    const workflow = {
      id: 12,
      name: 'Original Version',
      owner_name: 'John Johnson',
      public: true
    }

    const lesson = {
      course: {
        title: 'A Course',
        slug: 'a-course',
        localeId: 'en'
      },
      header: { title: 'A Lesson' }
    }

    const api = { duplicateWorkflow: jest.fn() }

    const wrapper = mountWithI18n(
      <WorkflowNavBar
        workflow={workflow}
        api={api}
        lesson={lesson}
        isReadOnly={false}
      /> // no loggedInUser prop
    )

    const a = wrapper.find('.course a')
    expect(a.prop('href')).toEqual('/courses/en/a-course')
    expect(a.text()).toEqual('A Course')
  })

  it('should duplicate the workflow when user is logged in and clicks the button', async () => {
    const workflow = {
      id: 12,
      name: 'Original Version',
      owner_name: 'John Johnson',
      public: true
    }

    const api = {
      duplicateWorkflow: jest.fn(() => Promise.resolve({
        id: 77,
        name: 'Copy of Original Version',
        owner_name: 'Paula Plagarizer',
        public: false
      }))
    }

    const wrapper = shallowWithI18n(
      <WorkflowNavBar
        workflow={workflow}
        api={api}
        isReadOnly={false}
        loggedInUser={{ id: 1 }}
      />
    )

    expect(wrapper).toMatchSnapshot()

    expect(wrapper.state().spinnerVisible).toBe(false)

    wrapper.find('button[name="duplicate"]').simulate('click')

    expect(api.duplicateWorkflow).toHaveBeenCalledWith(12)
    // spinner starts immediately
    expect(wrapper.state().spinnerVisible).toBe(true)
    // user isn't redirected yet
    expect(Utils.goToUrl).not.toHaveBeenCalled()

    await tick() // wait for promise to resolve

    expect(Utils.goToUrl).toHaveBeenCalledWith('/workflows/77')
  })

  it('should redirect to sign-in page when user clicks duplicate button while not logged in', () => {
    const workflow = {
      id: 303,
      name: 'Original Version',
      owner_name: 'Not LogggedIn',
      public: true
    }

    const api = { duplicateWorkflow: jest.fn() }

    const wrapper = shallowWithI18n(
      <WorkflowNavBar
        workflow={workflow}
        api={api}
        isReadOnly={false}
      /> // no loggedInUser prop
    )

    expect(wrapper).toMatchSnapshot()

    wrapper.find('button[name="duplicate"]').simulate('click')

    expect(api.duplicateWorkflow).not.toHaveBeenCalled()
    expect(Utils.goToUrl).toHaveBeenCalledWith('/account/login')
  })
})
