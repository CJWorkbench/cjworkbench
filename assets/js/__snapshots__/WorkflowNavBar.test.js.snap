// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkflowNavBar should duplicate the workflow when user is logged in and clicks the button 1`] = `
<Fragment>
  <nav
    className="navbar"
  >
    <div
      className="navbar-elements"
    >
      <a
        className="logo-navbar"
        href="/workflows/"
      >
        <img
          className="image"
          src="undefinedimages/logo.svg"
        />
      </a>
      <WorkflowTitleAndMetadata
        isReadOnly={false}
        workflow={
          Object {
            "id": 12,
            "name": "Original Version",
            "owner_name": "John Johnson",
            "public": true,
          }
        }
      />
      <div
        className="nav-buttons"
      >
        <WithI18n
          redo={[Function]}
          undo={[Function]}
        />
        <button
          name="duplicate"
          onClick={[Function]}
        >
          Duplicate
        </button>
        <WithI18n>
          Share
        </WithI18n>
        <WithI18n
          api={
            Object {
              "duplicateWorkflow": [MockFunction],
            }
          }
          isReadOnly={false}
          user={
            Object {
              "id": 1,
            }
          }
          workflowId={12}
        />
      </div>
    </div>
  </nav>
</Fragment>
`;

exports[`WorkflowNavBar should redirect to sign-in page when user clicks duplicate button while not logged in 1`] = `
<Fragment>
  <nav
    className="navbar"
  >
    <div
      className="navbar-elements"
    >
      <a
        className="logo-navbar"
        href="/workflows/"
      >
        <img
          className="image"
          src="undefinedimages/logo.svg"
        />
      </a>
      <WorkflowTitleAndMetadata
        isReadOnly={false}
        workflow={
          Object {
            "id": 303,
            "name": "Original Version",
            "owner_name": "Not LogggedIn",
            "public": true,
          }
        }
      />
      <div
        className="nav-buttons"
      >
        <WithI18n
          redo={[Function]}
          undo={[Function]}
        />
        <button
          name="duplicate"
          onClick={[Function]}
        >
          Duplicate
        </button>
        <WithI18n>
          Share
        </WithI18n>
        <a
          className="nav--link"
          href="/account/login"
        >
          Sign in
        </a>
      </div>
    </div>
  </nav>
</Fragment>
`;
